src/telemetry.erl

```
@@ -0,0 +1,136 @@
+    Handlers = telemetry_table_handler:list_for_event(EventName),
+    ApplyFun =
+        fun(#handler{handler_id=HandlerId,
+                     function=HandlerFunction,
+                     config=Config}) ->
+            try
+                HandlerFunction(EventName, EventValue, EventMetadata, Config)
+            catch
+                Class:Reason:Stacktrace ->
+                    detach(HandlerId),
+                    ?LOG_INFO("Handler ~p has failed and has been detached. "
+                              "Exception: class=~p reason=~p stacktrace=~p",
+                              [HandlerId, Class, Reason, Stacktrace])
+            end
+        end,
+
+    lists:foreach(ApplyFun, Handlers).
+
+
+%% @doc Returns all handlers attached to events with given prefix.
+%% Handlers attached to many events at once using `attach_many/4` will be listed once for each
+%% event they're attached to.
+%% Note that you can list all handlers by feeding this function an empty list.
+-spec list_handlers(event_prefix()) -> [handler()].
+list_handlers(EventPrefix) ->
+    assert_event_name_or_prefix(EventPrefix),
+    [#{handler_id => HandlerId,
+       event_name => EventName,
+       function => Function,
+       config => Config} || #handler{handler_id=HandlerId,
+                                     event_name=EventName,
+                                     function=Function,
+                                     config=Config} <- telemetry_table_handler:list_by_prefix(EventPrefix)].
+
+%%
+
+-spec assert_event_names_or_prefixes(term()) -> [ok].
+assert_event_names_or_prefixes(List) when is_list(List) ->
+    [assert_event_name_or_prefix(E) || E <- List].
+
+-spec assert_event_name_or_prefix(term()) -> ok.
+assert_event_name_or_prefix(List) when is_list(List) ->
+    case lists:all(fun erlang:is_atom/1, List) of
+        true ->
+            ok;
+        false ->
+            erlang:error(badarg)
```

Let's raise a `{badarg, {invalid_prefix, List}}` because `badarg` is not descriptive enough (Erlang raises both atoms and two-element tuple, so we should be fine).

---

Ah yup, meant to go back to that after asking since I wasn't sure what the equivalent to the Elixir raise stuff was. Sounds good.

---

Actually, I believe it should be `erlang:error(badarg, [List])`. That is what I'll update it to but can change it if you feel the `invalid_prefix` is important. The spec is `erlang:error(Reason, Args)`, then it can give the list of args in the stacktrace, and `invalid` seems kind of redundant to `badarg`.